art/defences/detector/evasion/binary_activation_detector.py

```diff
@@ -0,0 +1,107 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2023
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Module containing different methods for the detection of adversarial examples. All models are considered to be binary
+detectors.
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Tuple, Union, TYPE_CHECKING
+
+import numpy as np
+
+from art.defences.detector.evasion.evasion_detector import EvasionDetector
+
+if TYPE_CHECKING:
+    from art.estimators.classification.classifier import ClassifierNeuralNetwork
+
+logger = logging.getLogger(__name__)
+
+
+class BinaryActivationDetector(EvasionDetector):
+    """
+    Binary detector of adversarial samples coming from evasion attacks. The detector uses an architecture provided by
+    the user and is trained on the values of the activations of a classifier at a given layer.
+    """
+
+    defence_params = ["classifier", "detector", "layer"]
+
+    def __init__(
+        self,
+        classifier: "ClassifierNeuralNetwork",
+        detector: "ClassifierNeuralNetwork",
```

Comment:
Let's use `CLASSIFIER_NEURALNETWORK_TYPE` from `art.utils` which facilitates type checking:
```suggestion
        classifier: CLASSIFIER_NEURALNETWORK_TYPE,
        detector: CLASSIFIER_NEURALNETWORK_TYPE,
```